src/main/scala/is/hail/rvd/OrderedRVDPartitioner.scala

```diff
@@ -6,68 +6,129 @@ import is.hail.utils._
 
   val (pkType, _) = kType.select(partitionKey)
+  val pkIntervalType = TInterval(pkType)
+  val rangeBoundsType = TArray(pkIntervalType)
 
-  val pkKFieldIdx: Array[Int] = partitionKey.map(n => kType.fieldIdx(n))
-  val pkKOrd: UnsafeOrdering = OrderedRVDType.selectUnsafeOrdering(pkType, (0 until pkType.size).toArray, kType, pkKFieldIdx)
-
-  val rangeBoundsType = TArray(pkType)
   assert(rangeBoundsType.typeCheck(rangeBounds))
 
-  val ordering: Ordering[Annotation] = pkType.ordering.toOrdering
-  require(rangeBounds.isEmpty || rangeBounds.zip(rangeBounds.tail).forall { case (left, right) => ordering.compare(left, right) < 0 })
+  require(rangeBounds.isEmpty || (rangeBounds.zip(rangeBounds.tail).forall { case (left: Interval, right: Interval) =>
+    !left.overlaps(pkType.ordering, right) && pkType.ordering.lteq(left.start, right.start)
+  } && rangeBounds.forall { i => pkType.ordering.lteq(i.asInstanceOf[Interval].start, i.asInstanceOf[Interval].end) } ))
+
+  require(rangeBounds.isEmpty || rangeBounds.zip(rangeBounds.tail).forall { case (left: Interval, right: Interval) =>
+    pkType.ordering.equiv(left.end, right.start)})
+
+  val rangeTree: IntervalTree[Int] = IntervalTree.fromSorted(pkType.ordering,
+    Array.tabulate[(Interval, Int)](numPartitions) { i =>
+      (rangeBounds(i).asInstanceOf[Interval], i)
+    })
+
+  val pkKFieldIdx: Array[Int] = partitionKey.map(n => kType.fieldIdx(n))
 
   def region: Region = rangeBounds.region
 
-  def loadElement(i: Int): Long = rangeBoundsType.loadElement(rangeBounds.region, rangeBounds.aoff, rangeBounds.length, i)
+  def loadElement(i: Int): Long = rangeBoundsType.loadElement(region, rangeBounds.aoff, rangeBounds.length, i)
+
+  def loadStart(i: Int): Long = pkIntervalType.loadStart(region, loadElement(i))
+
+  def loadEnd(i: Int): Long = pkIntervalType.loadStart(region, loadElement(i))
+
 
-  // return the smallest partition for which key <= max
   // pk: Annotation[pkType]
-  def getPartitionPK(pk: Any): Int = {
+  // returns partition number if in partition, -1 otherwise
+  def checkPartitionPK(pk: Any): Int = {
     assert(pkType.typeCheck(pk))
+    val part = rangeTree.queryValues(pkType.ordering, pk)
+    part match {
+      case Array() => -1
+      case Array(x) => x
+    }
+  }
 
-    val part = BinarySearch.binarySearch(numPartitions,
-      // key.compare(elem)
-      i =>
-        if (i == numPartitions - 1)
-          -1 // key.compare(inf)
-        else
-          ordering.compare(pk, rangeBounds(i)))
-    part
+  // if outside bounds, throw error
+  // pk: Annotation[pkType]
+  def getPartitionPK(pk: Any): Int = {
+    assert(pkType.typeCheck(pk))
+    val part = rangeTree.queryValues(pkType.ordering, pk)
+    part match {
+      case Array(x) => x
+    }
   }
 
-  // return the smallest partition for which key <= max
+  // return the partition containing key
+  // if outside bounds, throw error
   // key: RegionValue[kType]
   def getPartition(key: Any): Int = {
```

Comment:
Having this fail on some keys is a change in the Spark behavior which I'm not 100% comfortable with because I'm not quite sure where the assumption is relied on.  I think all keys should be assigned to a partition (even if the partition's bounds don't contain the key), namely < min goes into the first partition and > max go into the last partition.

This will obviate the need for checkPartitionPK.